server/src/main/java/io/crate/planner/optimizer/rule/EquiJoinToLookupJoin.java

```diff
@@ -0,0 +1,216 @@
+            largerSide = rhsCollect;
+        } else {
+            smallerSide = rhsCollect;
+            largerSide = lhsCollect;
+        }
+
+        Symbol smallerRelationColumn = getOnlyElement(equiJoinCondition.get(smallerSide.relation().relationName()));
+        Symbol largerRelationColumn = getOnlyElement(equiJoinCondition.get(largerSide.relation().relationName()));
+
+        LogicalPlan lookupJoin = createLookup(
+            smallerSide,
+            smallerRelationColumn,
+            largerSide,
+            largerRelationColumn,
+            nodeCtx,
+            txnCtx
+        );
+
+        LogicalPlan newLhs;
+        LogicalPlan newRhs;
+
+        if (rhsIsLarger) {
+            newLhs = lhs;
+            newRhs = lookupJoin;
+        } else {
+            newLhs = lookupJoin;
+            newRhs = rhs;
+        }
+
+        return new JoinPlan(
+            newLhs,
+            newRhs,
+            plan.joinType(),
+            plan.joinCondition(),
+            plan.isFiltered(),
+            plan.isRewriteFilterOnOuterJoinToInnerJoinDone(),
+            true
+        );
+    }
+
+    private static LogicalPlan createLookup(Collect smallerRelation,
+                                            Symbol smallerRelationColumn,
+                                            Collect largerRelation,
+                                            Symbol largerRelationColumn,
+                                            NodeContext nodeCtx,
+                                            TransactionContext txnCtx) {
+
+        var lookUpQuery = new SelectSymbol(
+            new QueriedSelectRelation(
+                false,
+                List.of(smallerRelation.relation()),
+                List.of(),
+                List.of(smallerRelationColumn),
+                Literal.BOOLEAN_TRUE,
+                List.of(),
+                null,
+                null,
+                null,
+                null
+            ),
+            (ArrayType<?>) ArrayType.makeArray(smallerRelationColumn.valueType(), 1),
+            SelectSymbol.ResultType.SINGLE_COLUMN_MULTIPLE_VALUES,
+            true
+        );
+
+        var anyEqImplementation = nodeCtx.functions().get(
+            null,
+            AnyEqOperator.NAME,
+            List.of(largerRelationColumn, lookUpQuery),
+            txnCtx.sessionSettings().searchPath()
+        );
+
+        var anyEqFunction = new Function(
+            anyEqImplementation.signature(),
+            List.of(largerRelationColumn, lookUpQuery),
+            DataTypes.BOOLEAN
+        );
+        WhereClause whereClause = new WhereClause(anyEqFunction);
+        LogicalPlan largerSideWithLookup = new Collect(largerRelation.relation(), largerRelation.outputs(), whereClause);
+        LogicalPlan smallerSideIdLookup = new Collect(smallerRelation.relation(), List.of(smallerRelationColumn), WhereClause.MATCH_ALL);
+        smallerSideIdLookup = new RootRelationBoundary(smallerSideIdLookup);
```

Comment:
Couldn't a collect already have a where clause?

Reply:
yes true, good catch!

Reply:
Please also add a test case that would fail without the change